Wrappers/Python/cil/optimisation/functions/SVRGFunction.py

```diff
@@ -0,0 +1,236 @@
+        
+        Parameters
+        ----------
+        x : DataContainer
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+        function_num: `int` 
+            Between 0 and the number of functions in the list  
+        Returns
+        --------
+        DataContainer
+            the value of the approximate gradient of the sum function at :code:`x` given a `function_number` in {0,...,len(functions)-1}
+        """
+    
+
+        self._svrg_iter_number += 1
+
+        self.stoch_grad_at_iterate = self.functions[function_num].gradient(x)
+
+        if self.store_gradients is True:
+            self._stochastic_grad_difference = self.stoch_grad_at_iterate.sapyb(
+                1., self._list_stored_gradients[function_num], -1.)
+        else:
+            self._stochastic_grad_difference = self.stoch_grad_at_iterate.sapyb(
+                1., self.functions[function_num].gradient(self.snapshot), -1.)
+
+        self._update_data_passes_indices([function_num])
+
+        # full gradient is added to the stochastic grad difference
+        if out is None:
+            out = self._stochastic_grad_difference.sapyb(
+                self.num_functions, self._full_gradient_at_snapshot, 1.)
+        else:
+            self._stochastic_grad_difference.sapyb(
+                self.num_functions, self._full_gradient_at_snapshot, 1., out=out)
+
+        return out
+
+    def _update_full_gradient_and_return(self, x, out=None):
+        """
+        Updates the memory for full gradient computation. If :code:`store_gradients==True`, the gradient of all functions is computed and stored.
+        
+        Parameters
+        ----------
+        x : DataContainer
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+
+        Returns
+        --------
+        DataContainer
+            the value of the approximate gradient of the sum function at :code:`x` given a `function_number` in {0,...,len(functions)-1}
+        """
+
+        self._svrg_iter_number += 1
+
+        if self.store_gradients is True:
+            self._list_stored_gradients = [
+                fi.gradient(x) for fi in self.functions]
+            self._full_gradient_at_snapshot = sum(
+                self._list_stored_gradients, start=0*x)
+        else:
+            self._full_gradient_at_snapshot = self.full_gradient(x)
+            self.snapshot = x.copy()
+
+        self._update_data_passes_indices(list(range(self.num_functions)))
+
+        if out is None:
+            out = self._full_gradient_at_snapshot
+        else:
+            out.fill( self._full_gradient_at_snapshot)
+
+        return out
+
+
+class LSVRGFunction(SVRGFunction):
+    """""
+    A class representing a function for Loopless Stochastic Variance Reduced Gradient (SVRG) approximation. 
```

Comment:
I think essentially all the same comments from SVRG given above also apply to LSVRG, and in addition maybe state how the L version differs from the SVRG version. I won't add all these as individual comments below, so please check above.